server/src/main/java/org/opensearch/index/mapper/FlatObjectFieldMapper.java

```diff
@@ -0,0 +1,745 @@
+import java.util.Map;
+import java.util.function.BiFunction;
+import java.util.function.Supplier;
+
+import static org.opensearch.search.SearchService.ALLOW_EXPENSIVE_QUERIES;
+
+/**
+ * A field mapper for flat_objects. This mapper accepts JSON object and treat as string fields in one index.
+ * A flat_object field contains one parent field itself and two substring fields:
+ * field._valueAndPath and field._value
+ * @opensearch.internal
+ */
+public final class FlatObjectFieldMapper extends DynamicKeyFieldMapper {
+
+    public static final String CONTENT_TYPE = "flat_object";
+    private static final String VALUE_AND_PATH_SUFFIX = "._valueAndPath";
+    private static final String VALUE_SUFFIX = "._value";
+    private static final String DOT_SYMBOL = ".";
+    private static final String EQUAL_SYMBOL = "=";
+
+    /**
+     * In flat_object field mapper, field type is similar to keyword field type
+     * Cannot be tokenized, can OmitNorms, and can setIndexOption.
+     * @opensearch.internal
+     */
+    public static class Defaults {
+        public static final FieldType FIELD_TYPE = new FieldType();
+
+        static {
+            FIELD_TYPE.setTokenized(false);
+            FIELD_TYPE.setOmitNorms(true);
+            FIELD_TYPE.setIndexOptions(IndexOptions.DOCS);
+            FIELD_TYPE.freeze();
+        }
+
+    }
+
+    @Override
+    public MappedFieldType keyedFieldType(String key) {
+        return new FlatObjectFieldType(this.name() + DOT_SYMBOL + key);
+    }
+
+    /**
+     * FlatObjectFieldType is the parent field type.
+     */
+    public static class FlatObjectField extends Field {
+
+        public FlatObjectField(String field, BytesRef term, FieldType ft) {
+            super(field, term, ft);
+        }
+
+    }
+
+    private static FlatObjectFieldMapper toType(FieldMapper in) {
+        return (FlatObjectFieldMapper) in;
+    }
+
+    /**
+     * The builder for the flat_object field mapper using default parameters as
+     * indexed: flat_object field mapper is default to be indexed.
+     * hasDocValues: to store index and support efficient access to individual field values.
+     * stored: the original value of the field is not stored in the index.
+     * nullValue: not accept null value
+     * ignoreAbove: exclude values that exceed the maximum length from the indexing process.
+     * @opensearch.internal
+     */
+    public static class Builder extends FieldMapper.Builder<Builder> {
+
+        private final IndexAnalyzers indexAnalyzers;
+
+        Builder(String name, IndexAnalyzers indexAnalyzers) {
+            super(name, Defaults.FIELD_TYPE);
+            builder = this;
+            this.indexAnalyzers = indexAnalyzers;
+        }
+
+        public Builder(String name) {
+            this(name, null);
+        }
+
+        private FlatObjectFieldType buildFlatObjectFieldType(BuilderContext context, FieldType fieldType) {
+            NamedAnalyzer normalizer = Lucene.KEYWORD_ANALYZER;
```

Comment:
Didn't see this `normalizer ` being used in this method

Reply:
Why we not using  search analyzer and normalizer now but delaying it? I think it significantly limits the applicability of the `flat_object` field type, just simple lower/upper case queries would miss the search hits ...

Reply:
Currently, the FlatObjectFieldMapper is extended from DynamicKeyFieldMapper, which is extended from FieldMapper which does not parse parameters. I can see this as a limitation for now. But it's not a small changes to code piece to parse parameters.

In the earlier design, we haven't talked about the parameters, there will be more discussions later, but as we discussed before, we can plan on bringing parameters in the next PR. Assuming the first PR is using default parameters, we will explain in the documentation as well.

Reply:
use it in the next PR https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it